src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+    switch (split_type) {
+    case Serialize::SplitType::SplitType_SplitVar:
+        return Split::SplitType::SplitVar;
+    case Serialize::SplitType::SplitType_RenameVar:
+        return Split::SplitType::RenameVar;
+    case Serialize::SplitType::SplitType_FuseVars:
+        return Split::SplitType::FuseVars;
+    case Serialize::SplitType::SplitType_PurifyRVar:
+        return Split::SplitType::PurifyRVar;
+    default:
+        user_assert(false) << "unknown split type " << split_type << "\n";
+        return Split::SplitType::SplitVar;
+    }
+}
+
+DimType Deserializer::deserialize_dim_type(const Serialize::DimType dim_type) {
+    switch (dim_type) {
+    case Serialize::DimType::DimType_PureVar:
+        return DimType::PureVar;
+    case Serialize::DimType::DimType_PureRVar:
+        return DimType::PureRVar;
+    case Serialize::DimType::DimType_ImpureRVar:
+        return DimType::ImpureRVar;
+    default:
+        user_assert(false) << "unknown dim type " << dim_type << "\n";
+        return DimType::PureVar;
+    }
+}
+
+LoopAlignStrategy Deserializer::deserialize_loop_align_strategy(const Serialize::LoopAlignStrategy loop_align_strategy) {
+    switch (loop_align_strategy) {
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_AlignStart:
+        return LoopAlignStrategy::AlignStart;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_AlignEnd:
+        return LoopAlignStrategy::AlignEnd;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_NoAlign:
+        return LoopAlignStrategy::NoAlign;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_Auto:
+        return LoopAlignStrategy::Auto;
+    default:
+        user_assert(false) << "unknown loop align strategy " << loop_align_strategy << "\n";
+        return LoopAlignStrategy::AlignStart;
+    }
+}
+
+ExternFuncArgument::ArgType Deserializer::deserialize_extern_func_argument_type(const Serialize::ExternFuncArgumentType extern_func_argument_type) {
+    switch (extern_func_argument_type) {
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_UndefinedArg:
+        return ExternFuncArgument::ArgType::UndefinedArg;
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_FuncArg:
+        return ExternFuncArgument::ArgType::FuncArg;
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_BufferArg:
+        return ExternFuncArgument::ArgType::BufferArg;
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_ExprArg:
+        return ExternFuncArgument::ArgType::ExprArg;
+    case Serialize::ExternFuncArgumentType::ExternFuncArgumentType_ImageParamArg:
+        return ExternFuncArgument::ArgType::ImageParamArg;
+    default:
+        user_assert(false) << "unknown extern func argument type " << extern_func_argument_type << "\n";
+        return ExternFuncArgument::ArgType::UndefinedArg;
+    }
+}
+
+Type Deserializer::deserialize_type(const Serialize::Type *type) {
```

Comment:
Possibly-dumb question: if the argument can't legally be null, why not pass a reference instead of a pointer?

Reply:
some fields can possibly be null, for example, see `CreateParameter` usages in `Serialization.cpp`. For consistency across all deserialization function signatures, using pointers here seems more reasonable.